File: src/main/java/com/gitlab/daring/image/command/structure/GrubCutCenterCommand.java
package com.gitlab.daring.image.command.structure;

import com.gitlab.daring.image.command.BaseCommand;
import com.gitlab.daring.image.command.CommandEnv;
import com.gitlab.daring.image.command.parameter.IntParam;
import org.bytedeco.javacpp.opencv_core.*;

import static com.gitlab.daring.image.util.GeometryUtils.getCenterRect;
import static org.bytedeco.javacpp.opencv_core.*;
import static org.bytedeco.javacpp.opencv_imgproc.GC_INIT_WITH_RECT;
import static org.bytedeco.javacpp.opencv_imgproc.grabCut;

public class GrubCutCenterCommand extends BaseCommand {

	final IntParam size = intParam(0, 90, "0-100");
	final IntParam iterCount = intParam(1, 5, "0-100");

	final Mat bm = new Mat();
	final Mat fm = new Mat();

	public GrubCutCenterCommand(String... args) {
		super(args);
	}

	@Override
	public void execute(CommandEnv env) {
		Mat m = new Mat(env.mat.size(), CV_8U);
		Rect cr = getCenterRect(m.size(), size.v * 0.01);
		grabCut(env.mat, m, cr, bm, fm, iterCount.v, GC_INIT_WITH_RECT);
		env.mat = multiply(and(m, Scalar.all(1)), 255).asMat();
	}

}

File: src/main/kotlin/com/gitlab/daring/image/command/structure/GrubCutCenterCommand.kt
package com.gitlab.daring.image.command.structure

import com.gitlab.daring.image.command.CommandEnv
import com.gitlab.daring.image.command.KBaseCommand
import com.gitlab.daring.image.opencv.size
import org.bytedeco.javacpp.opencv_core.*
import org.bytedeco.javacpp.opencv_imgproc.*

internal class GrubCutCenterCommand(args: Array<String>) : KBaseCommand(args) {

    val r1 = intParam(5, "0-100")
    val r2 = intParam(90, "0-100")
    val mt = enumParam(MarkerType.Circle)
    val iterCount = intParam(5, "0-100")

    val rect = Rect()
    val bm = Mat()
    val fm = Mat()

    override fun execute(env: CommandEnv) {
        val m = Mat(env.mat.size, CV_8U, Scalar.BLACK)
        mt.v.drawCenter(m, r2, GC_PR_FGD, CV_FILLED)
        mt.v.drawCenter(m, r1, GC_FGD, CV_FILLED)
        //		env.mat = multiply(m, 128).asMat();
        grabCut(env.mat, m, rect, bm, fm, iterCount.v, GC_INIT_WITH_MASK)
        env.mat = multiply(and(m, Scalar.all(1.0)), 255.0).asMat()
    }

}
